devcareer/code-jammers-backend
# #174879632 Setup user account verification

## What does this PR do ##
- This PR enables a user to verify his/her account

### Description of tasks to be completed? ###
- Install necessary dependencies
- Send user an email to verify their account
- Verify user on visiting link provided in mail
- Setup sign-in verification to allow only verified users sign-in
- Test route to ensure user is verified successfully 

### How should this be manually tested/checked? ###
- Clone the PR
- Checkout to the branch `ft-user-verification-174879632`
- Install the dependencies using `npm install`
- Create a `.env file` and define the following as follows:
   - `DEV_DATABASE_URL = <development database url>`
   - `TEST_DATABASE_URL = <test database url>`
   - `DATABASE_URL = <production database url>`
   -  `JWT_KEY = secret`
   - `SENDGRID_API_KEY="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"`
   - `SENDGRID_EMAIL = "codejammers1@gmail.com"`
- Run migrations with the following command: `npm run migrate`
- Start the server by running `npm run start-dev`
- Sign-up a user using Postman with `http://localhost:3000/api/v1/users/signup` as URL.
  - For testing purposes you could create a temporary email with mailbox using this link `temp-mail.org`
  - request body JSON sample: 
```
{
    "username": "bellogo",
    "email": "belaogodo@gmail.com",
    "password": "password" 
}
```
- Sign-in with the same user using Postman with `http://localhost:3000/api/v1/users/signin` as url.
  - request body JSON sample: 
```
{
    "email": "belaogodo@gmail.com",
    "password": "password" 
}
```
- Login to user email and visit the verification link to verify user.
- Try Signing-in with same user using Postman with the sign-in url.


### What are the relevant pivotal tracker stories?
[#174879632](https://www.pivotaltracker.com/story/show/174879632)

## Architectural flowchart ##
[Verify User account flow chart](https://docs.google.com/document/d/13ALud7srMbKozKMBQLmh1HFBB95131tOI6_5kw7_5Nc/edit?usp=sharing)
